In [1]:
from IPython.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

In [2]:
# Import the modules
import numpy as np
import pandas as pd

from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv('Resources/lending_data.csv')

# Review the DataFrame
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts   
0    10700.0          7.672            52800        0.431818                5  \
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features

# Separate the y variable, the labels
Y = lending_data_df['loan_status']

# Separate the X variable, the features
X = lending_data_df.drop(columns='loan_status')

In [5]:
# Review the y variable Series
Y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [6]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts   
0    10700.0          7.672            52800        0.431818                5  \
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [8]:
# Check the balance of our target values
y_balance = Y.value_counts()
y_balance

loan_status
0    75036
1     2500
Name: count, dtype: int64

- Healthy loans (0): 75,036 instances
- At-risk loans (1): 2,500 instances

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [10]:
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, Y, random_state=1)

# Display the number of samples in each set
(len(X_train), len(X_test), len(y_train), len(y_test))

(58152, 19384, 58152, 19384)

#### The data has been split into training and testing datasets:

**Training data:**

- Features (X_train): 58,152 samples
- Labels (y_train): 58,152 samples

**Testing data:**

- Features (X_test): 19,384 samples
- Labels (y_test): 19,384 samples

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [11]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
logistic_regression_model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [12]:
# Make a prediction using the testing data
y_pred = logistic_regression_model.predict(X_test)

# Display the first few predictions
y_pred[:5]

array([0, 0, 0, 0, 0], dtype=int64)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [13]:
# Print the balanced_accuracy score of the model
balanced_acc_score = balanced_accuracy_score(y_test, y_pred)
balanced_acc_score

0.9520479254722232

In [14]:
# Generate a confusion matrix for the model
confusion_mat = confusion_matrix(y_test, y_pred)
confusion_mat

array([[18663,   102],
       [   56,   563]], dtype=int64)

In [15]:
# Print the classification report for the model
class_report = classification_report(y_test, y_pred)
class_report

'              precision    recall  f1-score   support\n\n           0       1.00      0.99      1.00     18765\n           1       0.85      0.91      0.88       619\n\n    accuracy                           0.99     19384\n   macro avg       0.92      0.95      0.94     19384\nweighted avg       0.99      0.99      0.99     19384\n'

### The logistic regression model's performance has been evaluated using various metrics, and here are the results:

1. Balanced Accuracy Score
The balanced accuracy score is approximately 
0.9521 or 95.21%.

2. Confusion Matrix
The confusion matrix is as follows:

    [18663,   102],
    [   56,   563]
Where:

- True Positive (TP): 563
- True Negative (TN): 18663
- False Positive (FP): 102 (Type I error)
- False Negative (FN): 56 (Type II error)

3. Classification Report
The classification report provides the following metrics:

**Precision:**

- Class 0 (Healthy loans): 1.00
- Class 1 (At-risk loans): 0.85

**Recall:**

- Class 0 (Healthy loans): 0.99
- Class 1 (At-risk loans): 0.91

**F1-score:**

- Class 0 (Healthy loans): 1.00
- Class 1 (At-risk loans): 0.88

**Support:**

- Class 0 (Healthy loans): 18765
- Class 1 (At-risk loans): 619

**Accuracy:** 0.99

**Macro Average:**

- Precision: 0.92
- Recall: 0.95
- F1-score: 0.94

**Weighted Average:**

- Precision: 0.99
- Recall: 0.99
- F1-score: 0.99

**Conclusion:**

The logistic regression model performed quite well on the testing data, with a high balanced accuracy score and high precision and recall values for both classes. The high F1-scores indicate a good balance between precision and recall.

### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 

The logistic regression model's performance in predicting both the 0 (healthy loan) and 1 (high-risk loan) labels can be assessed through various metrics as derived from the confusion matrix and classification report. Let's break down its performance for each label:

### **Healthy Loans (0)**
- **Precision:** The precision for predicting healthy loans is 1.00, indicating that almost all loans that the model predicted as healthy were indeed healthy. 
- **Recall:** The recall is 0.99, meaning that the model identified 99% of all actual healthy loans correctly.
- **F1-Score:** The F1-score, which is the harmonic mean of precision and recall, is 1.00, signifying a very balanced and excellent performance in predicting healthy loans.

### **High-Risk Loans (1)**
- **Precision:** The precision is 0.85, showing that 85% of the loans that the model labeled as high-risk were indeed high-risk.
- **Recall:** The recall is 0.91, meaning the model correctly identified 91% of all the actual high-risk loans in the dataset.
- **F1-Score:** The F1-score is 0.88, indicating a balanced and good performance in predicting high-risk loans, albeit slightly lower compared to the healthy loans prediction.

### **Overall Assessment**
- **Balanced Accuracy Score:** The balanced accuracy score is approximately 95.21%, which takes into account the true positive rate (sensitivity) and true negative rate (specificity), thus providing a balanced view of the model's performance across both classes.
- **Accuracy:** The overall accuracy is 99%, indicating that the model made correct predictions for 99% of the loans in the test dataset.

### **Conclusion**
- The logistic regression model performs exceptionally well in predicting healthy loans with high precision, recall, and F1-score. 
- It also performs quite well in predicting high-risk loans, although the precision and F1-score are slightly lower compared to those for healthy loans. 
- The high recall for high-risk loans indicates that the model is sensitive to detecting high-risk cases, which is often crucial in loan predictions to mitigate risk. 
- Thus, we can conclude that the logistic regression model predicts both healthy and high-risk loans effectively, with a slightly better performance for healthy loans.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [ ]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_oversampler = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = random_oversampler.fit_resample(X_train, y_train)

In [21]:
X_resampled.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts   
0     8600.0          6.792            44500        0.325843                3  \
1     7800.0          6.419            41000        0.268293                2   
2    10000.0          7.386            50100        0.401198                4   
3     9300.0          7.093            47300        0.365751                3   
4     9200.0          7.045            46900        0.360341                3   

   derogatory_marks  total_debt  
0                 0       14500  
1                 0       11000  
2                 1       20100  
3                 0       17300  
4                 0       16900

In [22]:
y_resampled.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [23]:
# Count the distinct values of the resampled labels data
y_resampled_balance = y_resampled.value_counts()
y_resampled_balance

loan_status
0    56271
1    56271
Name: count, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [24]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_resampled = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
logistic_regression_resampled.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
y_pred_resampled = logistic_regression_resampled.predict(X_test)

In [25]:
# Display the first few predictions
print(y_pred_resampled[:5])

[0 0 0 0 0]


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [29]:
# Print the balanced_accuracy score of the model 
balanced_acc_score_resampled = balanced_accuracy_score(y_test, y_pred_resampled)
balanced_acc_score_resampled

0.9936781215845847

In [27]:
# Generate a confusion matrix for the model
confusion_mat_resampled = confusion_matrix(y_test, y_pred_resampled)
confusion_mat_resampled

array([[18649,   116],
       [    4,   615]], dtype=int64)

In [28]:
# Print the classification report for the model
class_report_resampled = classification_report(y_test, y_pred_resampled)
class_report_resampled

'              precision    recall  f1-score   support\n\n           0       1.00      0.99      1.00     18765\n           1       0.84      0.99      0.91       619\n\n    accuracy                           0.99     19384\n   macro avg       0.92      0.99      0.95     19384\nweighted avg       0.99      0.99      0.99     19384\n'

### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 

Let's interpret the model's performance based on the balanced_accuracy score, a confusion matrix and classification report for the model:

### **1. Balanced Accuracy Score**
The balanced accuracy score is approximately \(0.9937\) or \(99.37\%\), which is an excellent score, indicating that the model has a high true positive rate and a high true negative rate.

### **2. Confusion Matrix**
The confusion matrix is as follows:
\[
\begin{array}{cc}
18649 & 116 \\
4 & 615 \\
\end{array}
\]
Where:
- **True Positive (TP):** 615
- **True Negative (TN):** 18649
- **False Positive (FP):** 116 (Type I error)
- **False Negative (FN):** 4 (Type II error)

### **3. Classification Report**
The classification report provides the following metrics:

#### **Healthy Loans (0)**

- **Precision:** 1.00
- **Recall:** 0.99
- **F1-score:** 1.00
- **Support:** 18765

#### **High-Risk Loans (1)**

- **Precision:** 0.84
- **Recall:** 0.99
- **F1-score:** 0.91
- **Support:** 619

#### **Overall**

- **Accuracy:** 0.99
- **Macro Average:**
  - **Precision:** 0.92
  - **Recall:** 0.99
  - **F1-score:** 0.95
- **Weighted Average:**
  - **Precision:** 0.99
  - **Recall:** 0.99
  - **F1-score:** 0.99

### **Conclusion**
The logistic regression model trained with oversampled data performs exceedingly well in predicting both healthy and high-risk loans:

- **Healthy Loans (0):** The model almost perfectly predicts the healthy loans, with very high precision, recall, and F1-score.
  
- **High-Risk Loans (1):** The model does extremely well in predicting high-risk loans, with a high recall of 0.99, meaning it correctly identified 99% of all the actual high-risk loans. The precision is slightly lower, indicating that some of the loans predicted as high-risk were actually healthy. However, the high F1-score of 0.91 indicates a good balance between precision and recall.

- **Overall:** The high balanced accuracy score and overall accuracy indicate that the model is highly effective in predicting both classes, with a slight tendency to misclassify some healthy loans as high-risk, which is a safer mistake in the context of loan predictions as it helps in avoiding potential bad loans.

Therefore, we can conclude that the logistic regression model, fit with oversampled data, predicts both the 0 (healthy loan) and 1 (high-risk loan) labels very effectively, demonstrating high sensitivity and specificity.